File: finder4j-backtrack-examples/src/main/java/com/github/K0zka/finder4j/backtrack/examples/sudoku/SudokuStep.java
package com.github.K0zka.finder4j.backtrack.examples.sudoku;

import java.util.Arrays;

import com.github.K0zka.finder4j.backtrack.Step;

@SuppressWarnings("serial")
public class SudokuStep implements Step<SudokuState> {

	public SudokuStep(short x, short y, short num) {
		super();
		this.x = x;
		this.y = y;
		this.num = num;
	}

	private final short num;
	private final short x;
	private final short y;

	public SudokuState take(SudokuState state) {
		final Short[][] copy = new Short[state.getSudoku().length][];
		for(int i = 0; i < state.getSudoku().length; i++) {
			copy[i] = Arrays.copyOf(state.getSudoku()[i], state.getSudoku().length);
		}
		copy[x][y] = num;
		return new SudokuState(copy);
	}

	public String toString() {
		return "{ [" + x + "," + y + "] <-" + num + " }";
	}
}

File: finder4j-backtrack-examples/src/main/kotlin/com/github/k0zka/finder4j/backtrack/examples/sudoku/SudokuStep.kt
package com.github.k0zka.finder4j.backtrack.examples.sudoku

import java.util.Arrays

import com.github.k0zka.finder4j.backtrack.Step

class SudokuStep(private val x: Short, private val y: Short, private val num: Short) : Step<SudokuState> {

	override fun take(state: SudokuState): SudokuState {
		val copy = Array<ShortArray>(state.sudoku.size,
										  {i -> Arrays.copyOf(state.sudoku[i], state.sudoku.size)}
		)
		for (i in 0 until state.sudoku.size) {
			copy[i] = state.sudoku[i].copyOf()
		}
		copy[x.toInt()][y.toInt()] = num
		return SudokuState(copy)
	}

	override fun toString(): String {
		return "{ [$x,$y] <-$num }"
	}
}
